### PR TITLE
Refactor HW models, widgets and how they are used in GraphEstimator

### PR DESCRIPTION
**Pull request by dexter2206**
## Description

Currently a hardware model is needed to instantiate `GraphEstimator`. What's more, during `GraphEstimator` initialization a `WidgetIterator` is constructed and stored in the estimator object. This makes `GraphEstimator` mutable in a very problematic way, as `WidgetIterator` can be iterated over exactly once. Thus, code like:

```python
result = estimator.estimate(algo)
result2 = estimator.estimate(algo)
```
will either fail or give incorrect results for `result2`. This PR addresses this shortcoming, by removing `WidgetIterator` entirely and instead relying on single dispatches to construct an iterator over matching widgets only when the iterator is needed. 

To facilitate the change, the hardware models are also altered. They are now implemented as distinct implementations of the `BasicArchitectureModel` protocol, so that functions creating widget iterables can dispatch on their type.

This PR also improves widget class by adding a `n_t_gates_produced` field (by default set to 1. This allows for removing conditional logic of the form:

```python
if "20-to-4" in widget.name:
    graph_data.n_nodes = ceil(graph_data.n_nodes / 4)
```

in favor of more adequate expressions like:
```python
graph_data.n_nodes = ceil(graph_data.n_nodes / widget.n_produced_t_gates)
```
which should facilitate further expansions of the list of available widgets.

Also, the ability to filter out "20-to-4" widgets is removed. Instead, factory that produces widget-iterables  in `GraphEstimator` can be customized, so one can easily replace the default one with one that filters 20-to-4 widgets if needed.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

File: src/benchq/data_structures/hardware_architecture_models.py [hunk 1]
```diff
@@ -24,9 +24,26 @@ class BasicArchitectureModel:
...
+
+
+@dataclass(frozen=True)
+class IONTrapModel:
+    physical_qubit_error_rate: float
+    surface_code_cycle_time_in_seconds: float
+
+
+@dataclass(frozen=True)
+class SCModel:
```
Comment (AthenaCaesura) [inline src/benchq/data_structures/hardware_architecture_models.py:43]: The this code allows for one to have an SCModel which has variable `physical_qubit_error_rate` and `surface_code_cycle_time_in_seconds`. We should only allow `physical_qubit_error_rate = 1e-4` and `surface_code_cycle_time_in_seconds = 1e-7` or else the widget selectiong gets messed up.

Comment (AthenaCaesura) [reply src/benchq/data_structures/hardware_architecture_models.py:43]: Same for the `IONTrapModel`.

File: src/benchq/resource_estimation/graph/graph_estimator.py [hunk 2]
```diff
@@ -84,13 +85,16 @@ def __init__(
         decoder_model: Optional[DecoderModel] = None,
         optimization: str = "space",
         substrate_scheduler_preset: str = "fast",
+        widget_iterable_factory: Callable[
+            [BasicArchitectureModel], Iterable[Widget]
+        ] = default_widget_list,
```
Comment (AthenaCaesura) [inline src/benchq/resource_estimation/graph/graph_estimator.py:90]: Is there any easier way to do this?? Factories for iterables only make the code much less readable...

File: src/benchq/resource_estimation/graph/graph_estimator.py [hunk 3]
```diff
@@ -286,29 +289,35 @@ def estimate_resources_from_graph_data(
...
-            for widget in self.widget_iterator:
-                if "20-to-4" in widget.name:
-                    old_n_nodes = graph_data.n_nodes
-                    graph_data.n_nodes = ceil(graph_data.n_nodes / 4)
+            widget_found = False
+            for widget in widget_iterator:
+                tmp_graph_data = replace(
+                    graph_data,
+                    n_nodes=ceil(graph_data.n_nodes / widget.n_t_gates_produced),
+                )
```
Comment (AthenaCaesura) [inline src/benchq/resource_estimation/graph/graph_estimator.py:302]: I love this! But can we change the name to `widget.n_t_gates_produced_per_cycle`?

**Review by AthenaCaesura — CHANGES_REQUESTED** [review]

File: src/benchq/data_structures/hardware_architecture_models.py (hunk 1 above)
Comment (SebastianMorawiec) [reply src/benchq/data_structures/hardware_architecture_models.py:43]: There are test usages of IONTrapModel and SCModel with different values than 1e-4 and 1e-7. Current approach suggests to use BASIC_* constant dataclasses for this reason.

File: src/benchq/resource_estimation/graph/graph_estimator.py (hunk 2 above)
Comment (SebastianMorawiec) [reply src/benchq/resource_estimation/graph/graph_estimator.py:90]: I can imagine this being complicated. Ive changed it to just Iterable[Widget] for anyone who wants to provide their own widget list, with default being chosen from default_widget_list based on hw_model.

**Review by AthenaCaesura — APPROVED** [review]

**Pull request merged**